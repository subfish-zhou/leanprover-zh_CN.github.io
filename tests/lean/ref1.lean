--
def inc (r : IO.Ref Nat) : IO Unit := do
let v ← r.get;
r.set (v+1);
IO.println (">> " ++ toString v)

def initArray (r : IO.Ref (Array Nat)) (n : Nat) : IO Unit :=
n.forM $ fun i => do
  r.modify $ fun a => a.push (2*i)

def showArrayRef (r : IO.Ref (Array Nat)) : IO Unit := do
let a ← r.swap ∅;
a.size.forM (fun i => IO.println ("[" ++ toString i ++ "]: " ++ toString (a.get! i)));
discard $ r.swap a;
pure ()

def tst (n : Nat) : IO Unit := do
let r₁ ← IO.mkRef 0;
n.forM fun _ => inc r₁;
let r₂ ← IO.mkRef (∅ : Array Nat);
initArray r₂ n;
showArrayRef r₂

#eval tst 10
