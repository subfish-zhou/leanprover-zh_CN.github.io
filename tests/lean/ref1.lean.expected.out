>> 0
>> 1
>> 2
>> 3
>> 4
>> 5
>> 6
>> 7
>> 8
>> 9
[0]: 0
[1]: 2
[2]: 4
[3]: 6
[4]: 8
[5]: 10
[6]: 12
[7]: 14
[8]: 16
[9]: 18
